src/runtime-rs/crates/hypervisor/src/dragonball/vmm_instance.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    fs::{File, OpenOptions},
+    os::unix::{io::IntoRawFd, prelude::AsRawFd},
+    sync::{
+        mpsc::{channel, Receiver, Sender},
+        Arc, Mutex, RwLock,
+    },
+    thread,
+};
+
+use anyhow::{anyhow, Context, Result};
+use dragonball::{
+    api::v1::{
+        BlockDeviceConfigInfo, BootSourceConfig, FsDeviceConfigInfo, FsMountConfigInfo,
+        InstanceInfo, InstanceState, VirtioNetDeviceConfigInfo, VmmAction, VmmActionError, VmmData,
+        VmmRequest, VmmResponse, VmmService, VsockDeviceConfigInfo,
+    },
+    vm::VmConfigInfo,
+    Vmm,
+};
+use nix::sched::{setns, CloneFlags};
+use seccompiler::BpfProgram;
+use vmm_sys_util::eventfd::EventFd;
+
+use crate::ShareFsOperation;
+
+pub enum Request {
+    Sync(VmmAction),
+}
+
+const DRAGONBALL_VERSION: &str = env!("CARGO_PKG_VERSION");
+const REQUEST_RETRY: u32 = 500;
+
+pub struct VmmInstance {
+    /// VMM instance info directly accessible from runtime
+    vmm_shared_info: Arc<RwLock<InstanceInfo>>,
+    to_vmm: Option<Sender<VmmRequest>>,
+    from_vmm: Option<Receiver<VmmResponse>>,
+    to_vmm_fd: EventFd,
+    seccomp: BpfProgram,
+    vmm_thread: Option<thread::JoinHandle<Result<i32>>>,
+}
+
+impl VmmInstance {
+    pub fn new(id: &str) -> Self {
+        let vmm_shared_info = Arc::new(RwLock::new(InstanceInfo::new(
+            String::from(id),
+            DRAGONBALL_VERSION.to_string(),
+        )));
+
+        let to_vmm_fd = EventFd::new(libc::EFD_NONBLOCK)
+            .unwrap_or_else(|_| panic!("Failed to create eventfd for vmm {}", id));
+
+        VmmInstance {
+            vmm_shared_info,
+            to_vmm: None,
+            from_vmm: None,
+            to_vmm_fd,
+            seccomp: vec![],
+            vmm_thread: None,
+        }
+    }
+
+    pub fn get_shared_info(&self) -> Arc<RwLock<InstanceInfo>> {
+        self.vmm_shared_info.clone()
+    }
+
+    fn set_instance_id(&mut self, id: &str) {
+        let share_info_lock = self.vmm_shared_info.clone();
+        share_info_lock.write().unwrap().id = String::from(id);
+    }
+
+    pub fn get_vcpu_tids(&self) -> Vec<(u8, u32)> {
+        let info = self.vmm_shared_info.clone();
+        let result = info.read().unwrap().tids.clone();
+        result
+    }
+
+    pub fn run_vmm_server(&mut self, id: &str, netns: Option<String>) -> Result<()> {
+        let kvm = OpenOptions::new().read(true).write(true).open("/dev/kvm")?;
```

Comment:
Might be worth making the device a parameter or a global variable to allow unit tests to mock this.

Reply:
Thanks for your review, change to use a global variable.